### PR TITLE
feature: Support `EXPLAIN COPY`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/pull/6539

## Rationale for this change

It is important to see what plans are coming out for copy

## What changes are included in this PR?

1. Add support for `EXPLAIN` plans

## Are these changes tested?
Yes

## Are there any user-facing changes?
`EXPLAIN` etc works now for `COPY` statements

### CODE REVIEW

File: datafusion/sql/src/parser.rs [hunk 1]
```diff
@@ -44,6 +44,35 @@ fn parse_file_type(s: &str) -> Result<String, ParserError> {
     Ok(s.to_uppercase())
 }
 
+/// DataFusion specific EXPLAIN (needed so we can EXPLAIN datafusion
```
Comment (alamb) [inline datafusion/sql/src/parser.rs:47]: This is the key thing -- we need a datafusion specific `EXPLAIN` statement

File: datafusion/sqllogictest/test_files/copy.slt [hunk 2]
```diff
@@ -25,12 +25,19 @@ COPY source_table TO 'test_files/scratch/table' (format parquet, per_thread_outp
...
-#Explain copy queries not currently working
-query error DataFusion error: This feature is not implemented: Unsupported SQL statement: Some\("COPY source_table TO 'test_files/scratch/table'"\)
+# Error case
+query error DataFusion error: Error during planning: Copy To format not explicitly set and unable to get file extension!
 EXPLAIN COPY source_table to 'test_files/scratch/table'
 
-query error DataFusion error: SQL error: ParserError\("Expected end of statement, found: source_table"\)
+query TT
 EXPLAIN COPY source_table to 'test_files/scratch/table' (format parquet, per_thread_output true)
+----
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/copy.slt:34]: Here is an example of the explain plan working

File: datafusion/sqllogictest/test_files/explain.slt [hunk 3]
```diff
@@ -101,13 +101,17 @@ set datafusion.explain.physical_plan_only = false
 
 
 ## explain nested
-statement error Explain must be root of the plan
+query error DataFusion error: Error during planning: Nested explain not supported
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/explain.slt:104]: I think these errors are a little more clear

File: datafusion/expr/src/logical_plan/plan.rs [hunk 4]
```diff
@@ -1100,15 +1100,13 @@ impl LogicalPlan {
...
                         for (key, val) in options {
                             if !op_str.is_empty() {
-                                op_str.push(',');
+                                op_str.push_str(", ");
                             }
                             op_str.push_str(&format!("{key} {val}"));
                         }
-                        op_str.push(')');
-                        write!(f, "CopyTo: format={file_format} output_url={output_url} per_thread_output={per_thread_output} options: {op_str}")
+                        write!(f, "CopyTo: format={file_format} output_url={output_url} per_thread_output={per_thread_output} options: ({op_str})")
```
[the remark this answers is not in the document]
Comment (devinjdangelo) [reply datafusion/expr/src/logical_plan/plan.rs:1109]: I also reworked this function in #7294 like so:
https://github.com/apache/arrow-datafusion/blob/429d0f0bb586bd9a20809faa4e3715e9586427f8/datafusion/expr/src/logical_plan/plan.rs#L1095-L1109

This version is less efficient as it loops over the values twice, but this is also not performance sensitive code so the tradeoff for readability maybe makes sense.

File: datafusion/sql/src/parser.rs (hunk 1 above)
Comment (devinjdangelo) [reply datafusion/sql/src/parser.rs:47]: Is there a reason to keep these datafusion specific statements in general vs. moving this upstream to the sqlparser crate? This distinction was difficult for me to understand when first diving into how statements are parsed.

Comment (alamb) [reply datafusion/sql/src/parser.rs:47]: I think the high level rationale is that sqlparser-rs intends to parse sql for one or more existing "standard" sql dialects (like MySQL or Postgres). However, these statements (`COPY` and `CREATE EXTERNAL TABLE` are DataFusion specific extensions). I will make a PR with some docs to to try and clarify this rationale

File: datafusion/expr/src/logical_plan/plan.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/expr/src/logical_plan/plan.rs]: I merged https://github.com/apache/arrow-datafusion/pull/7294 and updated this PR to use that code instead. Thanks @devinjdangelo

File: datafusion/sql/src/parser.rs [hunk 5]
```diff
@@ -213,11 +242,12 @@ impl fmt::Display for Statement {
             Statement::CreateExternalTable(stmt) => write!(f, "{stmt}"),
             Statement::DescribeTableStmt(_) => write!(f, "DESCRIBE TABLE ..."),
             Statement::CopyTo(stmt) => write!(f, "{stmt}"),
+            Statement::Explain(stmt) => write!(f, "{stmt}"),
         }
     }
 }
 
-/// DataFusion SQL Parser based on [`sqlparser`]
+/// Datafusion1 SQL Parser based on [`sqlparser`]
```
Comment (metesynnada) [inline datafusion/sql/src/parser.rs:250]:
```suggestion
/// Datafusion SQL Parser based on [`sqlparser`]
```

File: datafusion/sql/src/statement.rs [hunk 6]
```diff
@@ -706,13 +713,18 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
         &self,
         verbose: bool,
         analyze: bool,
-        statement: Statement,
+        statement: DFStatement,
     ) -> Result<LogicalPlan> {
-        let plan = self.sql_statement_to_plan(statement)?;
+        let plan = self.statement_to_plan(statement)?;
+        if matches!(plan, LogicalPlan::Explain(_)) {
+            return plan_err!("Nested explain not supported");
```
Comment (metesynnada) [inline datafusion/sql/src/statement.rs:726]: Maybe the error can be `Nested EXPLAINs are not supported.`.

**Review by metesynnada — COMMENTED** [review]
Looks good to me! Converting the hashmap into a vector was a clever move.

File: datafusion/sql/src/statement.rs (hunk 6 above)
Comment (alamb) [reply datafusion/sql/src/statement.rs]: 👍  in 5c46949ea

**Review by jackwener — APPROVED** [review]
Great job😍.

Next PR we can add it into doc.

https://arrow.apache.org/datafusion/user-guide/sql/explain.html

File: datafusion/sql/src/statement.rs [hunk 7]
```diff
@@ -706,13 +713,18 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
     ///
-    fn explain_statement_to_plan(
+    /// Note this is the sqlparser explain statement, not the
+    /// datafusion `EXPLAIN` statement.
+    fn explain_to_plan(
         &self,
         verbose: bool,
         analyze: bool,
-        statement: Statement,
+        statement: DFStatement,
```
Comment (jackwener) [inline datafusion/sql/src/statement.rs:722]: Can we use `ExplainStatement` as param?

Comment (alamb) [reply datafusion/sql/src/statement.rs:722]: So there are two places this is called -- one from the DataFusion `ExplainStatement` (introduced in this PR) and one from the SQL parser [`Explain`](https://docs.rs/sqlparser/0.36.1/sqlparser/ast/enum.Statement.html#variant.Explain). So I thought it best to support them both and simply pass the parameters on through that are actually used

**Pull request merged**